### PR TITLE
Full upwinding added to ConservativeAdvection

### LINKED ISSUE #11852: Enhance framework ConservativeAdvection with full upwinding option
## Rationale

The ConservativeAdvection Kernel is of limited use currently because it includes no numerical stabilisation.  I think it's really mainly a teaching tool on how to make MOOSE Kernels but wouldn't ever be used in any real simulations because of the crazy over/under shoots that it yields.  Including some stabilisation would be instructive for coders of MOOSE (they'd see nodal Values, not just normal _qp stuff) and would also allow this to be used in real simulations.  I want to include a full-upwinding option: someone else can do SUPG or something else if they really want to.

## Description

Make a MooseEnum that will let the user specify the upwinding type used: it defaults to "none"

For upwinding_type = none, just do what's currently in ConservativeAdvection

For upwinding_type = full, perform full upwinding.  That is, look at the velocity at each node.  Some nodes will have total outflow>0: evaluate the residual at these nodes based on their value of u.  Other nodes will have outflow<0: they get the flow from the outflow>0 nodes.

## Impact

As a teaching tool for how to do something slightly more complicated with Kernels, as well as making ConservativeAdvection more useful in real simulations.

### PR DESCRIPTION
**Pull request by WilkAndy**
Fixes #11852

### CODE REVIEW

File: framework/doc/content/source/kernels/ConservativeAdvection.md [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 ## Description
 
-The `ConservativeAdvection` kernel implements an advection term given for the domain ($\Omega$) define as
+The `ConservativeAdvection` kernel implements an advection term given for the domain ($\Omega$) defined as
 
 \begin{equation}
 \underbrace{\nabla \cdot \vec{v} u}_{\textrm{ConservativeAdvection}} + \sum_{i=1}^n \beta_i = 0 \in \Omega,
```
Comment (YaqiWang) [inline framework/doc/content/source/kernels/ConservativeAdvection.md]: beta term seems irrelevant in this documentation.

File: framework/doc/content/source/kernels/ConservativeAdvection.md [hunk 2]
```diff
@@ -22,27 +22,97 @@ element solution of the weak formulation. The first term is the volumetric term
...
-The corresponding Jacobian is given by
+Without numerical stabilization the corresponding Jacobian is given by
 
 \begin{equation}
 \frac{\partial R_i(u_h)}{\partial u_j} = -(\nabla \psi_i, \vec{v} \phi_j).
 \end{equation}
 
+## Full upwinding
+
+Advective flow is notoriously prone to physically-incorrect overshoots
```
Comment (YaqiWang) [inline framework/doc/content/source/kernels/ConservativeAdvection.md]: It appears to me that this scheme works only with continuous FEM, so maybe somewhere adding this restriction will be good.

File: framework/doc/content/source/kernels/ConservativeAdvection.md [hunk 3]
```diff
@@ -22,27 +22,97 @@ element solution of the weak formulation. The first term is the volumetric term
...
+
+## Dirichlet boundary conditions for pure advection
+
+In the purely advective case, Dirichlet boundary conditions may be
+used to inject mass (or heat, or whatever $u$ corresponds to
+physically) into the domain, or extract it from the domain.
+Specifically:
+
+- if $\vec{v}\cdot\vec{n}<0$ (with $\vec{n}$ being the outward normal) then the velocity field is "blowing" mass into the domain.  Fixing $u=u_{B}$ on this boundary means the mass flux (units kg.s$^{-1}$.m$^{-2}$) entering the domain is $-\vec{v}\cdot\vec{n}u_{B}$.
+- if $\vec{v}\cdot\vec{n}>0$ (with $\vec{n}$ being the outward normal) then the velocity field is "blowing" mass out of the domain.  If no Dirichlet boundary condition is used then mass is prevented from leaving the domain via this boundary: mass arriving here will "build up", causing $u$ to increase.  Alternatively, if $u=0$ is fixed at the boundary all mass arriving here will exit the domain immediately.
```
Comment (YaqiWang) [inline framework/doc/content/source/kernels/ConservativeAdvection.md]: Should we add an integrated BC on downwind: $\langle\psi_i, \vec{v} u
 \cdot \vec{n}\rangle$ (from the integration by part above)?

File: framework/doc/content/source/kernels/ConservativeAdvection.md [hunk 4]
```diff
@@ -22,27 +22,97 @@ element solution of the weak formulation. The first term is the volumetric term
...
-The velocity is supplied as a three component vector with order $v_x$, $v_y$, and  $v_z$.
+The velocity is supplied as a three component vector with order $v_x$,
+$v_y$, and  $v_z$.
+
+## Comparison of no upwinding and full upwinding
+
+The tests
+!listing test/tests/kernels/conservative_advection/no_upwinding_1D.i
+and
+!listing test/tests/kernels/conservative_advection/full_upwinding_1D.i
```
Comment (YaqiWang) [inline framework/doc/content/source/kernels/ConservativeAdvection.md]: I am curious: this problem will not reach stead state solution u=1?

File: framework/doc/content/source/kernels/ConservativeAdvection.md (hunk 1 above)
Comment (WilkAndy) [reply framework/doc/content/source/kernels/ConservativeAdvection.md]: This beta term was already in the previous documentation.  I had never thought of writing such a term before, but i like it.  With beta you get to specify the whole DE, rather than just the advection part, and it makes signs clearer.

File: framework/doc/content/source/kernels/ConservativeAdvection.md (hunk 2 above)
Comment (WilkAndy) [reply framework/doc/content/source/kernels/ConservativeAdvection.md]: True, thanks.  Done

File: framework/doc/content/source/kernels/ConservativeAdvection.md (hunk 3 above)
Comment (WilkAndy) [reply framework/doc/content/source/kernels/ConservativeAdvection.md]: Yes, you are probably correct.  i'll investigate next week.

Comment (WilkAndy) [reply framework/doc/content/source/kernels/ConservativeAdvection.md]: Actually, on second thoughts, i wonder whether this is necessary.  Fixing u at the boundary naturally does all the integrated stuff, doesn't it?

Comment (YaqiWang) [reply framework/doc/content/source/kernels/ConservativeAdvection.md]: So now we have three options: Dirichlet (specifying u), Integrated (using $\langle\psi_i, \vec{v} u
\cdot \vec{n}\rangle$), or none. I do not know how they are related.

Comment (YaqiWang) [reply framework/doc/content/source/kernels/ConservativeAdvection.md]: I typically use `IntegratedBC` to apply BC both on upwinding and downwinding. Just split $\langle\psi_i, \vec{v} u \cdot \vec{n}\rangle$ into two parts, for the upwinding, substitute u_B into u, for the dowdwinding, keep u. u=u_B will be a steady-state solution. And there is no discontinuity in the solution on the downwinding boundary like applying Dirichlet boundary there.

File: framework/doc/content/source/kernels/ConservativeAdvection.md (hunk 4 above)
Comment (YaqiWang) [reply framework/doc/content/source/kernels/ConservativeAdvection.md]: I see. lim(t->infty) u(x) = 1 on all x inside the domain except at x on the right boundary. Is du/dx at the right boundary equal to zero?

Comment (WilkAndy) [reply framework/doc/content/source/kernels/ConservativeAdvection.md]: In the limit t->infty, u(x)=1 for x<1, but u(1)=infty.

File: framework/doc/content/source/kernels/ConservativeAdvection.md (hunk 3 above)
Comment (WilkAndy) [reply framework/doc/content/source/kernels/ConservativeAdvection.md]: Here's my understanding of the BCs (also related to your comment below).  I'll rewrite the doco to reflect this, once you confirm that it agrees with your understanding.  Thanks @YaqiWang .   Basically, i think i was stupid in introducing DirichletBCs into the discussion.

To form the correct equations, we always need to include the <v.n u> in the equations (n is outward normal to the boundary).  In MOOSE we do this by always including a BC, which is called OutflowBC along every boundary.   Physically what this does is subtract away <v.n u> "fluid mass" from the boundary.

So for v.n>0 adding the OutflowBC allows "fluid" to flow freely through the boundary.  The velocity is "blowing fluid" into this boundary, and the OutflowBC is removing it at the correct rate, because the flux through any area is <v.n u>.

For v.n<0 adding the OutflowBC actually ends up fixing u at the boundary to its initial condition.  This is because the ConservativeAdvection Kernel is making fluid flow from the boundary to the interior of the model, but at the same time the OutflowBC is removing <v.n u> : note that v.n<0, so the OutflowBC is actually adding fluid at the same rate the ConservativeAdvection Kernel is adding it.  The physical interpretation is that something external to the model is adding fluid at exactly the rate specified by the initial conditions at that boundary.

Usually this isn't what we want for v.n<0.  Usually we want to specify a particular value for the injected flux, and we do that using InflowBC, which is just adding <v.n u_B>, where u_B (kg/m^2/s) is the injection rate.  (Using a DirichletBC to do this is just **wrong**, sorry.)

To make impermeable boundaries, either for v.n>0 or v.n<0 simply use no BC at that boundary.  Then for v.n>0 there is no BC to remove fluid from that boundary so the fluid "piles up" there.  For v.n<0 the omission of a BC can be thought of setting u_B=0 in an InflowBC.  (Again, using a DirichletBC is **wrong**, sorry).

File: framework/doc/content/source/kernels/ConservativeAdvection.md (hunk 4 above)
Comment (WilkAndy) [reply framework/doc/content/source/kernels/ConservativeAdvection.md]: Please see the extended comment above, @YaqiWang

File: framework/doc/content/source/kernels/ConservativeAdvection.md (hunk 3 above)
Comment (YaqiWang) [reply framework/doc/content/source/kernels/ConservativeAdvection.md]: Thanks Andy. Did you updated the doc to reflect your comment here?

File: framework/doc/content/source/kernels/ConservativeAdvection.md [hunk 5]
```diff
@@ -22,27 +22,104 @@ element solution of the weak formulation. The first term is the volumetric term
...
+Similarly, define
+\begin{equation}
+\tilde{M}_{\mathrm{in}} = - \sum_{i \ \mathrm{ with }\  \tilde{R}_{i}< 0}
+\tilde{R}_{i} \ .
+\end{equation}
+Mass is conserved if the residual for the downwind nodes is defined to
+be
+\begin{equation}
+R_{i} = \tilde{R}_{i} M_{\mathrm{out}} / M_{\mathrm{in}}  \ \ \texttt{ for } \ \ \tilde{R}_{i}< 0.
+\end{equation}
```
Comment (YaqiWang) [inline framework/doc/content/source/kernels/ConservativeAdvection.md]: I should have asked this: this treatment works with the case where other terms exist like the diffusion or source term, right? I knew SUPG cannot be done just on one term. Maybe a small paragraph on discussing the advantage of this scheme will answer readers' questions like this.

Comment (WilkAndy) [reply framework/doc/content/source/kernels/ConservativeAdvection.md]: Yes, you may use full upwinding on just the advective term if you have other terms (like diffusion).  It is a problem-specific choice, really.  If your diffusion (etc) terms are complicated nonlinear things you may choose to fully-upwind them too but you don't *need* to.

Comment (WilkAndy) [reply framework/doc/content/source/kernels/ConservativeAdvection.md]: Addressed: added a paragraph to doco.

**Comment by YaqiWang** [comment]
This looks very good to me. I guess I have one more question: is this treatment consistent, i.e. refining the mesh will lead the numerical solution to the analytical solution unbiased?

**Comment by WilkAndy** [comment]
Yea, you get convergence to the correct solution as you refine your mesh.  The main problems are:
 - increased diffusion
 - problems achieving steady-state if the velocity swaps directions every nonlinear iteration (not the case for ConservativeAdvection, of course).

I added this paragraph:

Full upwinding adds more numerical diffusion than most other numerical
stabilization techniques such as SUPG and TVD.  Another problem is
that steady-state can be hard to achieve in nonlinear problems where
the velocity is not fixed and changes every nonlinear iteration (this
does not occur for `ConservativeAdvection`).  On the other hand, full
upwinding is computationally cheap.

**Comment by YaqiWang** [comment]
Sorry, @WilkAndy I just looked `Kernel.C`, I think you should follow the code pattern in it, for example, use `accumulateTaggedLocalResidual` instead of doing the accumulation in the old way. One of MOOSE guys @permcody should review and merge this.

**Comment by WilkAndy** [comment]
Yep, i agree about the accumulate, @YaqiWang .   i'll do that in the next few hours,

a

File: framework/doc/content/source/kernels/ConservativeAdvection.md [hunk 6]
```diff
@@ -20,29 +20,154 @@ R_i(u_h) = \underbrace{-(\nabla \psi_i, \vec{v} u)}_{\textrm{ConservativeAdvecti
...
+Full-upwinding [cite:dalen1979,adhikary2011] is an example of
+numerical stabilization and this essentially adds numerical diffusion
+to completely eliminate overshoots and undershoots.  Full-upwinding is
+available in `ConservativeAdvection` by setting the `upwinding_type`
+appropriately.
+
+!alert note
+Full upwinding only works for continuous FEM
+
+In DE systems describing more than just advection (eg, in
```
Comment (andrsd) [inline framework/doc/content/source/kernels/ConservativeAdvection.md]: `s/eg/e.g./`

File: framework/doc/content/source/kernels/ConservativeAdvection.md [hunk 7]
```diff
@@ -20,29 +20,154 @@ R_i(u_h) = \underbrace{-(\nabla \psi_i, \vec{v} u)}_{\textrm{ConservativeAdvecti
...
 
-The velocity is supplied as a three component vector with order $v_x$, $v_y$, and  $v_z$.
+The velocity is supplied as a three component vector with order $v_x$,
+$v_y$, and  $v_z$.
+
+## Boundary conditions for pure advection
+
+To form the correct equations, the boundary term $\langle\psi_i, \vec{v} u
+\cdot \vec{n}\rangle$ needs to be included in the `BCs` block of a
+MOOSE input file (n is the outward normal to the boundary).  An
```
Comment (andrsd) [inline framework/doc/content/source/kernels/ConservativeAdvection.md]: `s/n/$n$/`, similarly like you do it for $v_x$ above

File: framework/doc/content/source/kernels/ConservativeAdvection.md [hunk 8]
```diff
@@ -20,29 +20,154 @@ R_i(u_h) = \underbrace{-(\nabla \psi_i, \vec{v} u)}_{\textrm{ConservativeAdvecti
...
+rate.  For instance
+
+!listing test/tests/kernels/conservative_advection/full_upwinding_1D.i block=BCs
+
+To make impermeable boundaries, either for $\vec{v} \cdot \vec{n} < 0$
+or $\vec{v} \cdot \vec{n} > 0$, simply use
+no BC at that boundary. Then for $\vec{v} \cdot \vec{n} > 0$ there is no BC to remove fluid
+from that boundary so the fluid "piles up" there. For $\vec{v} \cdot \vec{n} < 0$ the
+omission of a BC can be thought of setting $u_B=0$ in an
+InflowBC.
```
Comment (andrsd) [inline framework/doc/content/source/kernels/ConservativeAdvection.md]: s/InflowBC/\`InflowBC\`/

File: framework/doc/content/source/kernels/ConservativeAdvection.md [hunk 9]
```diff
@@ -20,29 +20,154 @@ R_i(u_h) = \underbrace{-(\nabla \psi_i, \vec{v} u)}_{\textrm{ConservativeAdvecti
...
+
+and
+
+!listing test/tests/kernels/conservative_advection/full_upwinding_1D.i
+
+describe the same physical situation: advection from left to right
+along a line.  A source at the left boundary introduces $u$ into the
+domain at a fixed rate (using an `InflowBC`).  The right boundary is
+impermeable (no BC), so when $u$ arrives
+there is starts building up at the boundary.  It is clear from the
```
Comment (andrsd) [inline framework/doc/content/source/kernels/ConservativeAdvection.md]: `s/is/it/`

**Review by andrsd — CHANGES_REQUESTED** [review]

**Review by lindsayad — COMMENTED** [review]
This is nice work, and I think I'm fine with it being in the framework. But I know we've had some fierce debates before on the complexity of objects that we allow in. This would probably become the most complex computing object in the framework proper if it's merged. I think I would lean towards allowing complexity to increase the general utility of the object, especially since it defaults to simplicity (through the `MooseEnum`) and there is nothing physics specific here. @friedmud probably has an opinion on this

File: framework/doc/content/source/kernels/ConservativeAdvection.md (hunk 6 above)
Comment (WilkAndy) [reply framework/doc/content/source/kernels/ConservativeAdvection.md]: Done

File: framework/doc/content/source/kernels/ConservativeAdvection.md (hunk 7 above)
Comment (WilkAndy) [reply framework/doc/content/source/kernels/ConservativeAdvection.md]: Done

File: framework/doc/content/source/kernels/ConservativeAdvection.md (hunk 8 above)
Comment (WilkAndy) [reply framework/doc/content/source/kernels/ConservativeAdvection.md]: Done

File: framework/doc/content/source/kernels/ConservativeAdvection.md (hunk 9 above)
Comment (WilkAndy) [reply framework/doc/content/source/kernels/ConservativeAdvection.md]: Done

**Comment by WilkAndy** [comment]
I addressed @andrsd requests, and will push up soon.

@lindsayad , i think i gave reasonable reasons why this should be in the framework in Issue #11852 .

**Review by lindsayad — APPROVED** [review]

**Review by andrsd — APPROVED** [review]

**Comment by WilkAndy** [comment]
Give @friedmud another day...

File: framework/src/kernels/ConservativeAdvection.C [hunk 10]
```diff
@@ -19,22 +19,169 @@ validParams<ConservativeAdvection>()
...
+  for (_i = 0; _i < num_nodes; ++_i)
+  {
+    for (_qp = 0; _qp < _qrule->n_points(); _qp++)
+      _local_re(_i) += _JxW[_qp] * _coord[_qp] * negSpeedQp();
+    _upwind_node[_i] = (_local_re(_i) >= 0.0);
+  }
+
+  // Variables used to ensure mass conservation
+  Real total_mass_out = 0.0;
+  Real total_in = 0.0;
```
Comment (dschwen) [inline framework/src/kernels/ConservativeAdvection.C]: `total_mass_in` maybe?

**Comment by WilkAndy** [comment]
Let's merge!
